File: docx_parsers/styles/run_properties_parser.py
# run_properties_parser.py

import xml.etree.ElementTree as ET
from typing import Optional
from docx_parsers.helpers.common_helpers import extract_element, extract_attribute, extract_boolean_attribute
from docx_parsers.utils import convert_half_points_to_points
from docx_parsers.models.styles_models import RunStyleProperties, FontProperties, LanguageProperties

class RunPropertiesParser:
    """
    A parser for extracting run properties from an XML element.
    """

    def parse(self, rPr_element: ET.Element) -> RunStyleProperties:
        """
        Parses run properties from the given XML element.

        Args:
            rPr_element (ET.Element): The run properties XML element.

        Returns:
            RunStyleProperties: The parsed run style properties.
        """
        properties = RunStyleProperties()

        if rPr_element is not None:
            properties.font = self.extract_fonts(rPr_element)
            properties.size_pt = self.extract_font_size(rPr_element)
            properties.color = self.extract_font_color(rPr_element)
            properties.bold = self.extract_bold(rPr_element)
            properties.italic = self.extract_italic(rPr_element)
            properties.underline = self.extract_underline(rPr_element)
            properties.strikethrough = self.extract_strikethrough(rPr_element)
            properties.hidden = self.extract_hidden(rPr_element)
            properties.lang = self.extract_language_settings(rPr_element)
            properties.highlight = self.extract_highlight(rPr_element)
            properties.shading = self.extract_shading(rPr_element)
            properties.text_position_pt = self.extract_text_position(rPr_element)
            properties.kerning = self.extract_kerning(rPr_element)
            properties.character_spacing_pt = self.extract_character_spacing(rPr_element)
            properties.emboss = self.extract_emboss(rPr_element)
            properties.outline = self.extract_outline(rPr_element)
            properties.shadow = self.extract_shadow(rPr_element)
            properties.all_caps = self.extract_all_caps(rPr_element)
            properties.small_caps = self.extract_small_caps(rPr_element)

        return properties

    def extract_fonts(self, rPr_element: ET.Element) -> Optional[FontProperties]:
        font_element = extract_element(rPr_element, "w:rFonts")
        if font_element is not None:
            font_properties = FontProperties()
            font_properties.ascii = extract_attribute(font_element, 'ascii')
            font_properties.hAnsi = extract_attribute(font_element, 'hAnsi')
            font_properties.eastAsia = extract_attribute(font_element, 'eastAsia')
            font_properties.cs = extract_attribute(font_element, 'cs')
            return font_properties
        return None

    def extract_font_size(self, rPr_element: ET.Element) -> Optional[float]:
        size_element = extract_element(rPr_element, "w:sz")
        if size_element is not None:
            size = extract_attribute(size_element, 'val')
            if size:
                return convert_half_points_to_points(int(size))
        return None

    def extract_font_color(self, rPr_element: ET.Element) -> Optional[str]:
        color_element = extract_element(rPr_element, "w:color")
        if color_element is not None:
            return extract_attribute(color_element, 'val')
        return None

    def extract_bold(self, rPr_element: ET.Element) -> Optional[bool]:
        bold_element = extract_element(rPr_element, "w:b")
        return extract_boolean_attribute(bold_element)

    def extract_italic(self, rPr_element: ET.Element) -> Optional[bool]:
        italic_element = extract_element(rPr_element, "w:i")
        return extract_boolean_attribute(italic_element)

    def extract_underline(self, rPr_element: ET.Element) -> Optional[str]:
        underline_element = extract_element(rPr_element, "w:u")
        if underline_element is not None:
            return extract_attribute(underline_element, 'val')
        return None

    def extract_strikethrough(self, rPr_element: ET.Element) -> Optional[bool]:
        strikethrough_element = extract_element(rPr_element, "w:strike")
        return extract_boolean_attribute(strikethrough_element)

    def extract_hidden(self, rPr_element: ET.Element) -> Optional[bool]:
        hidden_element = extract_element(rPr_element, "w:vanish")
        return extract_boolean_attribute(hidden_element)

    def extract_language_settings(self, rPr_element: ET.Element) -> Optional[LanguageProperties]:
        lang_element = extract_element(rPr_element, "w:lang")
        if lang_element is not None:
            lang_properties = LanguageProperties()
            lang_properties.val = extract_attribute(lang_element, 'val')
            lang_properties.eastAsia = extract_attribute(lang_element, 'eastAsia')
            lang_properties.bidi = extract_attribute(lang_element, 'bidi')
            return lang_properties
        return None

    def extract_highlight(self, rPr_element: ET.Element) -> Optional[str]:
        highlight_element = extract_element(rPr_element, "w:highlight")
        if highlight_element is not None:
            return extract_attribute(highlight_element, 'val')
        return None

    def extract_shading(self, rPr_element: ET.Element) -> Optional[str]:
        shading_element = extract_element(rPr_element, "w:shd")
        if shading_element is not None:
            return extract_attribute(shading_element, 'val')
        return None

    def extract_text_position(self, rPr_element: ET.Element) -> Optional[float]:
        text_position_element = extract_element(rPr_element, "w:position")
        if text_position_element is not None:
            text_position = extract_attribute(text_position_element, 'val')
            if text_position:
                return convert_half_points_to_points(int(text_position))
        return None

    def extract_kerning(self, rPr_element: ET.Element) -> Optional[int]:
        kerning_element = extract_element(rPr_element, "w:kern")
        if kerning_element is not None:
            kerning = extract_attribute(kerning_element, 'val')
            if kerning:
                return int(kerning)
        return None

    def extract_character_spacing(self, rPr_element: ET.Element) -> Optional[float]:
        character_spacing_element = extract_element(rPr_element, "w:spacing")
        if character_spacing_element is not None:
            character_spacing = extract_attribute(character_spacing_element, 'val')
            if character_spacing:
                return convert_half_points_to_points(int(character_spacing))
        return None

    def extract_emboss(self, rPr_element: ET.Element) -> Optional[bool]:
        emboss_element = extract_element(rPr_element, "w:emboss")
        return extract_boolean_attribute(emboss_element)

    def extract_outline(self, rPr_element: ET.Element) -> Optional[bool]:
        outline_element = extract_element(rPr_element, "w:outline")
        return extract_boolean_attribute(outline_element)

    def extract_shadow(self, rPr_element: ET.Element) -> Optional[bool]:
        shadow_element = extract_element(rPr_element, "w:shadow")
        return extract_boolean_attribute(shadow_element)

    def extract_all_caps(self, rPr_element: ET.Element) -> Optional[bool]:
        all_caps_element = extract_element(rPr_element, "w:caps")
        return extract_boolean_attribute(all_caps_element)

    def extract_small_caps(self, rPr_element: ET.Element) -> Optional[bool]:
        small_caps_element = extract_element(rPr_element, "w:smallCaps")
        return extract_boolean_attribute(small_caps_element)

File: docx_parser_converter_ts/src/docx_parsers/styles/runPropertiesParser.ts
import type {
  RunStyleProperties,
  FontProperties,
  LanguageProperties,
} from '../models/stylesModels';
import {
  extractElement,
  extractAttribute,
  extractBooleanAttribute,
  safeInt,
} from '../helpers/commonHelpers';
import { convertHalfPointsToPoints } from '../utils';

/**
 * A parser for extracting run properties from an XML element.
 *
 * This class extracts and parses various properties related to run formatting,
 * converting them into structured TypeScript interfaces for further processing or
 * conversion to other formats.
 */
export class RunPropertiesParser {
  /**
   * Parses run properties from the given XML element.
   *
   * @param rPrElement - The run properties XML element.
   * @returns The parsed run style properties.
   *
   * @example
   * The following is an example of run properties in a run properties element:
   * ```xml
   * <w:rPr>
   *     <w:rFonts w:ascii="Calibri" w:hAnsi="Calibri"/>
   *     <w:sz w:val="24"/>
   *     <w:color w:val="FF0000"/>
   *     <w:b/>
   *     <w:i/>
   *     <w:u w:val="single"/>
   *     <w:strike/>
   *     <w:highlight w:val="yellow"/>
   *     <w:shd w:val="clear" w:color="auto" w:fill="FFFF00"/>
   *     <w:position w:val="2"/>
   *     <w:kern w:val="2"/>
   *     <w:spacing w:val="2"/>
   *     <w:emboss/>
   *     <w:outline/>
   *     <w:shadow/>
   *     <w:caps/>
   *     <w:smallCaps/>
   * </w:rPr>
   * ```
   */
  public parse(rPrElement: Element | null): RunStyleProperties {
    const properties: RunStyleProperties = {};

    if (rPrElement) {
      properties.font = this.extractFonts(rPrElement);
      properties.sizePt = this.extractFontSize(rPrElement);
      properties.color = this.extractFontColor(rPrElement);
      properties.bold = this.extractBold(rPrElement);
      properties.italic = this.extractItalic(rPrElement);
      properties.underline = this.extractUnderline(rPrElement);
      properties.strikethrough = this.extractStrikethrough(rPrElement);
      properties.hidden = this.extractHidden(rPrElement);
      properties.lang = this.extractLanguageSettings(rPrElement);
      properties.highlight = this.extractHighlight(rPrElement);
      properties.shading = this.extractShading(rPrElement);
      properties.textPositionPt = this.extractTextPosition(rPrElement);
      properties.kerning = this.extractKerning(rPrElement);
      properties.characterSpacingPt = this.extractCharacterSpacing(rPrElement);
      properties.emboss = this.extractEmboss(rPrElement);
      properties.outline = this.extractOutline(rPrElement);
      properties.shadow = this.extractShadow(rPrElement);
      properties.allCaps = this.extractAllCaps(rPrElement);
      properties.smallCaps = this.extractSmallCaps(rPrElement);
    }

    return properties;
  }

  /**
   * Extracts font properties from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted font properties, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:rFonts w:ascii="Calibri" w:hAnsi="Calibri" w:eastAsia="SimSun" w:cs="Arial"/>
   * ```
   */
  private extractFonts(rPrElement: Element): FontProperties | undefined {
    const fontElement = extractElement(rPrElement, ".//w:rFonts");
    if (fontElement) {
      const fontProperties: FontProperties = {};
      const ascii = extractAttribute(fontElement, 'ascii');
      const hAnsi = extractAttribute(fontElement, 'hAnsi');
      const eastAsia = extractAttribute(fontElement, 'eastAsia');
      const cs = extractAttribute(fontElement, 'cs');

      if (ascii) fontProperties.ascii = ascii;
      if (hAnsi) fontProperties.hAnsi = hAnsi;
      if (eastAsia) fontProperties.eastAsia = eastAsia;
      if (cs) fontProperties.cs = cs;

      return Object.keys(fontProperties).length > 0 ? fontProperties : undefined;
    }
    return undefined;
  }

  /**
   * Extracts font size from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted font size in points, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:sz w:val="24"/>
   * ```
   */
  private extractFontSize(rPrElement: Element): number | undefined {
    const sizeElement = extractElement(rPrElement, ".//w:sz");
    if (sizeElement) {
      const size = extractAttribute(sizeElement, 'val');
      const sizeInt = safeInt(size);
      if (sizeInt !== null) {
        return convertHalfPointsToPoints(sizeInt);
      }
    }
    return undefined;
  }

  /**
   * Extracts font color from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted font color, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:color w:val="FF0000"/>
   * ```
   */
  private extractFontColor(rPrElement: Element): string | undefined {
    const colorElement = extractElement(rPrElement, ".//w:color");
    if (colorElement) {
      const color = extractAttribute(colorElement, 'val');
      return color ?? undefined;
    }
    return undefined;
  }

  /**
   * Extracts bold property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted bold property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:b/>
   * ```
   */
  private extractBold(rPrElement: Element): boolean | undefined {
    const boldElement = extractElement(rPrElement, ".//w:b");
    const val = extractBooleanAttribute(boldElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts italic property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted italic property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:i/>
   * ```
   */
  private extractItalic(rPrElement: Element): boolean | undefined {
    const italicElement = extractElement(rPrElement, ".//w:i");
    const val = extractBooleanAttribute(italicElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts underline property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted underline property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:u w:val="single"/>
   * ```
   */
  private extractUnderline(rPrElement: Element): string | undefined {
    const underlineElement = extractElement(rPrElement, ".//w:u");
    if (underlineElement) {
      return extractAttribute(underlineElement, 'val') ?? undefined;
    }
    return undefined;
  }

  /**
   * Extracts strikethrough property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted strikethrough property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:strike/>
   * ```
   */
  private extractStrikethrough(rPrElement: Element): boolean | undefined {
    const strikethroughElement = extractElement(rPrElement, ".//w:strike");
    const val = extractBooleanAttribute(strikethroughElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts hidden property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted hidden property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:vanish/>
   * ```
   */
  private extractHidden(rPrElement: Element): boolean | undefined {
    const hiddenElement = extractElement(rPrElement, ".//w:vanish");
    const val = extractBooleanAttribute(hiddenElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts language settings from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted language properties, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:lang w:val="en-US" w:eastAsia="zh-CN" w:bidi="ar-SA"/>
   * ```
   */
  private extractLanguageSettings(rPrElement: Element): LanguageProperties | undefined {
    const langElement = extractElement(rPrElement, ".//w:lang");
    if (langElement) {
      const langProperties: LanguageProperties = {};
      const val = extractAttribute(langElement, 'val');
      const eastAsia = extractAttribute(langElement, 'eastAsia');
      const bidi = extractAttribute(langElement, 'bidi');

      if (val) langProperties.val = val;
      if (eastAsia) langProperties.eastAsia = eastAsia;
      if (bidi) langProperties.bidi = bidi;

      return Object.keys(langProperties).length > 0 ? langProperties : undefined;
    }
    return undefined;
  }

  /**
   * Extracts highlight property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted highlight property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:highlight w:val="yellow"/>
   * ```
   */
  private extractHighlight(rPrElement: Element): string | undefined {
    const highlightElement = extractElement(rPrElement, ".//w:highlight");
    if (highlightElement) {
      return extractAttribute(highlightElement, 'val') ?? undefined;
    }
    return undefined;
  }

  /**
   * Extracts shading property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted shading property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:shd w:val="clear" w:color="auto" w:fill="FFFF00"/>
   * ```
   */
  private extractShading(rPrElement: Element): string | undefined {
    const shadingElement = extractElement(rPrElement, ".//w:shd");
    if (shadingElement) {
      return extractAttribute(shadingElement, 'val') ?? undefined;
    }
    return undefined;
  }

  /**
   * Extracts text position property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted text position in points, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:position w:val="2"/>
   * ```
   */
  private extractTextPosition(rPrElement: Element): number | undefined {
    const textPositionElement = extractElement(rPrElement, ".//w:position");
    if (textPositionElement) {
      const position = extractAttribute(textPositionElement, 'val');
      const positionInt = safeInt(position);
      if (positionInt !== null) {
        return convertHalfPointsToPoints(positionInt);
      }
    }
    return undefined;
  }

  /**
   * Extracts kerning property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted kerning property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:kern w:val="2"/>
   * ```
   */
  private extractKerning(rPrElement: Element): number | undefined {
    const kerningElement = extractElement(rPrElement, ".//w:kern");
    if (kerningElement) {
      const kerning = extractAttribute(kerningElement, 'val');
      const kerningInt = safeInt(kerning);
      return kerningInt ?? undefined;
    }
    return undefined;
  }

  /**
   * Extracts character spacing property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted character spacing in points, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:spacing w:val="2"/>
   * ```
   */
  private extractCharacterSpacing(rPrElement: Element): number | undefined {
    const spacingElement = extractElement(rPrElement, ".//w:spacing");
    if (spacingElement) {
      const spacing = extractAttribute(spacingElement, 'val');
      const spacingInt = safeInt(spacing);
      if (spacingInt !== null) {
        return convertHalfPointsToPoints(spacingInt);
      }
    }
    return undefined;
  }

  /**
   * Extracts emboss property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted emboss property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:emboss/>
   * ```
   */
  private extractEmboss(rPrElement: Element): boolean | undefined {
    const embossElement = extractElement(rPrElement, ".//w:emboss");
    const val = extractBooleanAttribute(embossElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts outline property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted outline property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:outline/>
   * ```
   */
  private extractOutline(rPrElement: Element): boolean | undefined {
    const outlineElement = extractElement(rPrElement, ".//w:outline");
    const val = extractBooleanAttribute(outlineElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts shadow property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted shadow property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:shadow/>
   * ```
   */
  private extractShadow(rPrElement: Element): boolean | undefined {
    const shadowElement = extractElement(rPrElement, ".//w:shadow");
    const val = extractBooleanAttribute(shadowElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts all caps property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted all caps property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:caps/>
   * ```
   */
  private extractAllCaps(rPrElement: Element): boolean | undefined {
    const allCapsElement = extractElement(rPrElement, ".//w:caps");
    const val = extractBooleanAttribute(allCapsElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts small caps property from the given run properties element.
   *
   * @param rPrElement - The run properties element.
   * @returns The extracted small caps property, or undefined if not present.
   *
   * @example
   * ```xml
   * <w:smallCaps/>
   * ```
   */
  private extractSmallCaps(rPrElement: Element): boolean | undefined {
    const smallCapsElement = extractElement(rPrElement, ".//w:smallCaps");
    const val = extractBooleanAttribute(smallCapsElement);
    return val === null ? undefined : val;
  }
} 